<a href="https://colab.research.google.com/github/vedantdave77/Natural_Laguage_Processing/blob/master/ML_Pipeline(evaluation_Comparison_bet_RF_and_GB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Building Machine Learning Classifiers: Model selection

### Read in & clean text

In [1]:
import nltk
nltk.download('stopwords')
import pandas as pd
import re
from sklearn.feature_extraction.text import TfidfVectorizer
import string

stopwords = nltk.corpus.stopwords.words('english')
ps = nltk.PorterStemmer()

data = pd.read_csv("SMSSpamCollection.tsv", sep='\t')
data.columns = ['label', 'body_text']

def count_punct(text):
    count = sum([1 for char in text if char in string.punctuation])
    return round(count/(len(text) - text.count(" ")), 3)*100

data['body_len'] = data['body_text'].apply(lambda x: len(x) - x.count(" "))
data['punct%'] = data['body_text'].apply(lambda x: count_punct(x))

def clean_text(text):
    text = "".join([word.lower() for word in text if word not in string.punctuation])
    tokens = re.split('\W+', text)
    text = [ps.stem(word) for word in tokens if word not in stopwords]
    return text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


### Split into train/test

In [0]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data[['body_text', 'body_len', 'punct%']], data['label'], test_size=0.2)

### Vectorize text

Use for change word to numbers for runing problem faster. and  (make bag to worlds) , 

In [3]:
tfidf_vect = TfidfVectorizer(analyzer=clean_text)
tfidf_vect_fit = tfidf_vect.fit(X_train['body_text'])

tfidf_train = tfidf_vect_fit.transform(X_train['body_text'])  # train and test has same data features not same data OK!.
tfidf_test = tfidf_vect_fit.transform(X_test['body_text'])

X_train_vect = pd.concat([X_train[['body_len', 'punct%']].reset_index(drop=True), 
           pd.DataFrame(tfidf_train.toarray())], axis=1)
X_test_vect = pd.concat([X_test[['body_len', 'punct%']].reset_index(drop=True), 
           pd.DataFrame(tfidf_test.toarray())], axis=1)

X_train_vect.head()

body_len  punct%         0    1    2  ...  7158  7159  7160  7161  7162
0        59     5.1  0.000000  0.0  0.0  ...   0.0   0.0   0.0   0.0   0.0
1        24     4.2  0.000000  0.0  0.0  ...   0.0   0.0   0.0   0.0   0.0
2        45     4.4  0.000000  0.0  0.0  ...   0.0   0.0   0.0   0.0   0.0
3        39     2.6  0.228176  0.0  0.0  ...   0.0   0.0   0.0   0.0   0.0
4        27    14.8  0.000000  0.0  0.0  ...   0.0   0.0   0.0   0.0   0.0

[5 rows x 7165 columns]

### Final evaluation of models

In [0]:
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import precision_recall_fscore_support as score
import time

In [5]:
rf = RandomForestClassifier(n_estimators=150, max_depth=None, n_jobs=-1)

rf_model = rf.fit(X_train_vect, y_train)
y_pred = rf_model.predict(X_test_vect)

precision, recall, fscore, train_support = score(y_test, y_pred, pos_label='spam', average='binary')
print('Precision: {} / Recall: {} / Accuracy: {}'.format(
    round(precision, 3), round(recall, 3), round((y_pred==y_test).sum()/len(y_pred), 3)))

Precision: 1.0 / Recall: 0.806 / Accuracy: 0.975


In [6]:
gb = GradientBoostingClassifier(n_estimators=150, max_depth=11)

gb_model = gb.fit(X_train_vect, y_train)
y_pred = gb_model.predict(X_test_vect)

precision, recall, fscore, train_support = score(y_test, y_pred, pos_label='spam', average='binary')
print('Precision: {} / Recall: {} / Accuracy: {}'.format(
    round(precision, 3), round(recall, 3), round((y_pred==y_test).sum()/len(y_pred), 3)))

Precision: 0.886 / Recall: 0.812 / Accuracy: 0.962


### Time calculation and final evaluation by comparing between RandomForest and gtadient booster for various parameters

In [7]:
gb = GradientBoostingClassifier(n_estimators=150, max_depth=11)

start = time.time()
gb_model = gb.fit(X_train_vect, y_train)
end = time.time()
fit_time = (end - start)

start = time.time()
y_pred = gb_model.predict(X_test_vect)
end = time.time()
pred_time = (end - start)

precision, recall, fscore, train_support = score(y_test, y_pred, pos_label='spam', average='binary')
print('Fit time: {} / Predict time: {} ---- Precision: {} / Recall: {} / Accuracy: {}'.format(
    round(fit_time, 3), round(pred_time, 3), round(precision, 3), round(recall, 3), round((y_pred==y_test).sum()/len(y_pred), 3)))

Fit time: 245.01 / Predict time: 0.086 ---- Precision: 0.908 / Recall: 0.819 / Accuracy: 0.966


In [8]:
rf = RandomForestClassifier(n_estimators=150, max_depth=11)

start = time.time()
rf_model = rf.fit(X_train_vect, y_train)
end = time.time()
fit_time = (end - start)

start = time.time()
y_pred = rf_model.predict(X_test_vect)
end = time.time()
pred_time = (end - start)

precision, recall, fscore, train_support = score(y_test, y_pred, pos_label='spam', average='binary')
print('Fit time: {} / Predict time: {} ---- Precision: {} / Recall: {} / Accuracy: {}'.format(
    round(fit_time, 3), round(pred_time, 3), round(precision, 3), round(recall, 3), round((y_pred==y_test).sum()/len(y_pred), 3)))

Fit time: 1.883 / Predict time: 0.063 ---- Precision: 1.0 / Recall: 0.382 / Accuracy: 0.92


### Gradient Booster
Fit time: 245.01 / Predict time: 0.086 ---- Precision: 0.908 / Recall: 0.819 / Accuracy: 0.966

### Random Forest
Fit time: 1.883 / Predict time: 0.063 ---- Precision: 1.0 / Recall: 0.382 / Accuracy: 0.92

### Observation
here for same depth and same no of estimators , we get higher accuracy in the Gradient booster but time for model training and prediction is higher than randomforestclassifier. So, if the accuracy does not have major difference then better to go with easy and light timed model to train and test the data.